src/Week5/HW05.hs

```diff
@@ -0,0 +1,144 @@
+{-# OPTIONS_GHC -Wall #-}
+{-# LANGUAGE OverloadedStrings, RecordWildCards #-}
+module Week5.HW05 where
+
+import Data.List
+import Data.ByteString.Lazy (ByteString)
+import Data.Map.Strict (Map)
+import System.Environment (getArgs)
+
+import qualified Data.ByteString.Lazy as BS
+import qualified Data.Map.Strict as Map
+
+import Data.Bits ( Bits(xor) )
+
+import Week5.Parser
+
+-- Exercise 1 -----------------------------------------
+
+getSecret :: FilePath -> FilePath -> IO ByteString
+getSecret path1 path2 =
+  do
+    f1 <- BS.readFile path1
+    f2 <- BS.readFile path2
+    return $ BS.filter (/=0) $ BS.pack $ BS.zipWith xor f1 f2
+
+
+
+-- Exercise 2 -----------------------------------------
+
+decryptWithKey :: ByteString -> FilePath -> IO ()
+decryptWithKey key output =
+  do
+    input <- BS.readFile $ output ++ ".enc"
+    BS.writeFile output $ BS.pack $ BS.zipWith xor input $ BS.cycle key
+
+
+-- Exercise 3 -----------------------------------------
+
+parseFile :: FromJSON a => FilePath -> IO (Maybe a)
+parseFile fp = do
+  f <- BS.readFile fp
+  return $ decode f
+
+-- Exercise 4 -----------------------------------------
+
+getBadTs :: FilePath -> FilePath -> IO (Maybe [Transaction])
+getBadTs victimPath transactionPath =
+  do
+    vf <- parseFile victimPath
+    tf <- parseFile transactionPath
+    let res = case (vf, tf) of 
+          (Nothing , _) -> Nothing 
+          (_, Nothing ) -> Nothing 
+          (Just v, Just t) -> Just $ helper v t 
+    return res
+      where 
+        helper :: [TId] -> [Transaction] -> [Transaction]
+        helper victim transactions = filter ((`elem` victim) . tid) transactions
+
+
+
+
+-- Exercise 5 -----------------------------------------
+
+getFlow :: [Transaction] -> Map String Integer
+getFlow ts = helper ts Map.empty 
+  where 
+    helper :: [Transaction] -> Map String Integer -> Map String Integer
+    helper [] mp = mp 
+    helper (x:xs) mp = 
+      let amt = amount x in 
+        helper xs $ Map.insertWith (+) (from x) amt $ Map.insertWith (+) (to x) (negate amt) mp
+
+-- Exercise 6 -----------------------------------------
+
+getCriminal :: Map String Integer -> String
+getCriminal mp = helper (Map.toList mp) 0 "all is good"
+  where 
+    helper :: [(String, Integer)] -> Integer -> String -> String
+    helper [] _ crim = crim
```

Comment:
(optional extra challenge to try)
You could use list foldr to make this simpler.
But even better would be not convert map to list to do the processing, but process the map directly using 
[foldrWithKey](https://hackage.haskell.org/package/containers-0.6.4.1/docs/Data-Map-Strict.html#v:foldrWithKey) or similar